File: cpp/problem_0021/merge_two_sorted_lists.cpp
/*
 * [21] Merge Two Sorted Lists
 *
 * You are given the heads of two sorted linked lists list1 and list2.

Merge the two lists into one sorted list. The list should be made by splicing together the nodes of
the first two lists.

Return the head of the merged linked list.


Example 1:


Input: list1 = [1,2,4], list2 = [1,3,4]
Output: [1,1,2,3,4,4]


Example 2:


Input: list1 = [], list2 = []
Output: []


Example 3:


Input: list1 = [], list2 = [0]
Output: [0]



Constraints:


    The number of nodes in both lists is in the range [0, 50].
    -100 <= Node.val <= 100
    Both list1 and list2 are sorted in non-decreasing order.
 *
 * Difficulty: Easy
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
  public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        ListNode *l3 = new ListNode();
        ListNode *head = l3;

        while (l1 && l2) {
            if (l1->val <= l2->val) {
                l3->next = l1;
                l1 = l1->next;
            } else {
                l3->next = l2;
                l2 = l2->next;
            }

            l3 = l3->next;
        }

        if (l1 != nullptr) {
            l3->next = l1;
        }

        if (l2 != nullptr) {
            l3->next = l2;
        }

        return head->next;
    }
};

File: typescript/problem_0021/MergeTwoSortedLists.ts
/*
 * [21] Merge Two Sorted Lists
 *
 * Difficulty: TODO
 */

const mergeTwoLists = function (l1: ListNode | null, l2: ListNode | null): ListNode | null {
  let head = new ListNode(-1);
  let l3 = head;
  while (l1 && l2) {
    if (l1.val >= l2.val) {
      l3.next = l2;
      l2 = l2.next;
    } else {
      l3.next = l1;
      l1 = l1.next;
    }
    l3 = l3.next;
  }

  if (l1) {
    l3.next = l1;
  }

  if (l2) {
    l3.next = l2;
  }
  return head.next;
};

export { mergeTwoLists };
